main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -1351,30 +1351,68 @@ object Resolver {
           }
 
         case NamedAst.Expr.StructNew(name, fields, region, loc) =>
-          val fieldsVal = traverse(fields) {
-            case (f, e) =>
-              val eVal = visitExp(e, env0)
-              mapN(eVal) {
-                case e => (f, e)
+          lookupStruct(name, env0, ns0, root) match {
+
+            case Result.Ok(st0) =>
+              flatMapN(getStructIfAccessible(st0, ns0, loc)) {
+                case st =>
+                  val providedFieldNames = fields.map {case (k, _) => k.name}.toSet
+                  val expectedFieldNames = st.fields.map(_.name.name).toSet
+                  val extraFields = providedFieldNames.diff(expectedFieldNames)
+                  val unprovidedFields = expectedFieldNames.diff(providedFieldNames)
+
+                  val errors = extraFields.map(ResolutionError.ExtraStructField(_, loc)) ++ unprovidedFields.map(ResolutionError.UnprovidedStructField(_, loc))
+                  val fieldsVal = traverse(fields) {
+                    case (f, e) =>
+                      val eVal = visitExp(e, env0)
+                      mapN(eVal) {
+                        case e => (f, e)
+                      }
+                  }
+                  val regionVal = visitExp(region, env0)
+                  val structNew = mapN(fieldsVal, regionVal) {
+                    case (fields, region) =>
+                      ResolvedAst.Expr.StructNew(st.sym, fields, region, loc)
+                  }
+                  structNew.withSoftFailures(errors)
               }
-          }
-          val regionVal = visitExp(region, env0)
-          mapN(fieldsVal, regionVal) {
-            case (fields, region) =>
-              throw new RuntimeException("Joe TBD")
+            case Result.Err(e) =>
+              Validation.toSoftFailure(ResolvedAst.Expr.Error(e), e)
           }
 
-        case NamedAst.Expr.StructGet(sym, e, field, loc) =>
-          val eVal = visitExp(e, env0)
-          mapN(eVal) {
-            throw new RuntimeException("Joe TBD")
+        case NamedAst.Expr.StructGet(name, e, field, loc) =>
+          lookupStruct(name, env0, ns0, root) match {
+            case Result.Ok(st) =>
+              if(!st.fields.map(_.name.name).contains(field.name)) {
```

Comment:
I think it would be useful with some named local variables, e.g.:

availableFields
usedFields

or whatever

Reply:
same comment applies elsewhere.